### PR TITLE
Failed keys implementation

### PR DESCRIPTION
**Pull request by vyeshwanth**
Adds support for failed keys. 

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: router/failed-events-manager.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+		for _, failedEvent := range failedEvents {
+			if len(failedEvent.RecordID) == 0 || !json.Valid(failedEvent.RecordID) {
+				pkgLogger.Infof("skipped adding invalid recordId: %s, to failed keys table: %s", failedEvent.RecordID, table)
+				continue
+			}
+			_, err = stmt.Exec(failedEvent.DestinationID, failedEvent.RecordID, createdAt)
+			if err != nil {
+				panic(err)
+			}
+		}
```
Comment (RanjeetMishra) [inline router/failed-events-manager.go:83]: minor: Would a batch update query be efficient? If so can we track as backlog task?

File: router/failed-events-manager.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+			if diff > failedKeysExpire {
+				dropQuery := fmt.Sprintf(`DROP TABLE IF EXISTS %s`, table)
+				rows, err = dbHandle.Query(dropQuery)
+				if err != nil {
+					pkgLogger.Errorf("Failed to drop table %s with error: %v", table, err)
+				}
+			}
+		}
+		dbHandle.Close()
+		time.Sleep(failedKeysCleanUpSleep)
```
Comment (RanjeetMishra) [inline router/failed-events-manager.go:175]: minor: Not sure whats a better option. An ever running process with sleep vs using a scheduled timer with registered function.

File: gateway/gateway.go [hunk 3]
```diff
@@ -1251,6 +1320,9 @@ func (gateway *HandleT) StartWebHandler(ctx context.Context) error {
 
 	//todo: remove in next release
 	srvMux.HandleFunc("/v1/pending-events", gateway.stat(gateway.pendingEventsHandler)).Methods("POST")
+	srvMux.HandleFunc("/v1/clear", gateway.stat(gateway.ClearHandler)).Methods("POST")
```
[the remark this answers is not in the document]
Comment (vyeshwanth) [reply gateway/gateway.go:1323]: this endpoint clears jobs in gateway, process and router tables for given, sourceId, destinationId and jobRunID. 
the implementation for this added before.

File: processor/processor.go [hunk 4]
```diff
@@ -1587,6 +1602,18 @@ func (proc *HandleT) processJobsForDest(jobList []*jobsdb.JobT, parsedEventList
...
 }
 
+func (proc *HandleT) saveFailedJobs(failedJobs []*jobsdb.JobT) {
+	if len(failedJobs) > 0 {
+		txn := proc.errorDB.BeginGlobalTransaction()
+		jobRunIDAbortedEventsMap := make(map[string][]*router.FailedEventRowT)
+		for _, failedJob := range failedJobs {
+			router.SaveSourceFailedEvents(failedJob.Parameters, jobRunIDAbortedEventsMap)
+		}
+		router.GetFailedEventsManager().SaveFailedRecordIDs(jobRunIDAbortedEventsMap, txn)
```
[the remark this answers is not in the document]
Comment (vyeshwanth) [reply processor/processor.go]: updated

**Review by RanjeetMishra — APPROVED** [review]
LGTM. Minor comments we will be following up as backlog task.

File: router/failed-events-manager.go (hunk 1 above)
Comment (vyeshwanth) [reply router/failed-events-manager.go:83]: Added a task
https://www.notion.so/rudderstacks/cfe8e8ef60a34485a373e789b7b6933a?v=a07a0b1350f84d46af9792260b39a194&p=278a8bbf4c1546aa942d6cf171dd168a

File: router/failed-events-manager.go (hunk 2 above)
Comment (vyeshwanth) [reply router/failed-events-manager.go:175]: Added a task in backlog
https://www.notion.so/rudderstacks/cfe8e8ef60a34485a373e789b7b6933a?v=a07a0b1350f84d46af9792260b39a194&p=a91b5dcdfe414780bb7bd7648c9d4c05

**Review by chandumlg — APPROVED** [review]

**Pull request merged**